File: Readers/MZXmlRawData.cpp
// Written by Navdeep Jaitly for the Department of Energy (PNNL, Richland, WA)
// Copyright 2006, Battelle Memorial Institute
// E-mail: navdeep.jaitly@pnl.gov
// Website: http://ncrr.pnl.gov/software
// -------------------------------------------------------------------------------
// 
// Licensed under the Apache License, Version 2.0; you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at 
// http://www.apache.org/licenses/LICENSE-2.0

#include ".\mzxmlrawdata.h"
#include "float.h"
#include <iostream> 
namespace Engine
{
	namespace Readers
	{
		MZXmlRawData::MZXmlRawData(void)
		{
			mint_num_scans = 0 ; 
			mptr_FILE = NULL ; 
			mptr_ScanIndex = NULL ; 
			mint_current_scan = -1 ; 
		}

		MZXmlRawData::~MZXmlRawData(void)
		{
			if (mptr_FILE != NULL)
			{
				Close() ;
				mptr_FILE = NULL ;
			}
			if (mptr_ScanIndex != NULL)
				delete [] mptr_ScanIndex ;
		}

		const char* MZXmlRawData::GetFileName()
		{
			return mstr_file_name ; 
		}

		FileType MZXmlRawData::GetFileType()
		{
			return MZXMLRAWDATA ; 
		}
		
		void MZXmlRawData::Close()
		{
			mint_num_scans = 0 ; 
			if (mptr_FILE == NULL)
				return ; 
			fclose(mptr_FILE) ; 
			mptr_FILE = NULL ; 
		}

		int MZXmlRawData::GetNumScans()
		{
			return mint_num_scans ; 
		}

		double MZXmlRawData::GetScanTime(int scan_num)
		{
			if (mint_current_scan == scan_num)
			{
				return mobj_CurrentScanHeader.retentionTime ; 
			}
			else
			{
				struct ScanHeaderStruct scanHeader ;

				// its time to read in that scan.
				mint_current_scan = scan_num ; 
				readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &scanHeader) ;
				return scanHeader.retentionTime ; 
			}
		}

		int MZXmlRawData::GetScanSize()
		{
			return 0 ; 
		}

		void MZXmlRawData::Load (char *file_name)
		{
			strcpy(mstr_file_name, file_name) ; 
			if (mptr_FILE != NULL)
			{
				delete mptr_FILE ; 
				mptr_FILE = NULL ; 
			}
			mptr_FILE = fopen(mstr_file_name, "r") ; 
			if (mptr_FILE == NULL)
			{
				char err_msg[MAX_ERR_LEN] ; 
				strcpy(err_msg, "Unable to open MZXML file: ") ; 
				strcat(err_msg, mstr_file_name) ; 
				throw err_msg ; 
			}

			if (mptr_ScanIndex != NULL)
			{
				delete [] mptr_ScanIndex ;
			}

			mpos_IndexOffset = getIndexOffset(mptr_FILE) ;
			readIndex(mptr_FILE , mpos_IndexOffset, &mint_num_scans, mptr_ScanIndex);
		}

		int MZXmlRawData::GetParentScan(int scan_num)
		{
			int scan = scan_num;
			struct ScanHeaderStruct scanHeader ; 
			readHeader(mptr_FILE, mptr_ScanIndex[scan], &scanHeader) ;	

			for (scan = scan_num - 1;; scan--)
			{
				readHeader(mptr_FILE, mptr_ScanIndex[scan], &scanHeader);	
				if (scanHeader.msLevel == 1)
					break;
			}

			return scan ; 
		}

		bool MZXmlRawData::IsFTScan(int scan_num)
		{
			int scan = scan_num ; 			
			int error = -1 ; 
			struct ScanHeaderStruct scanHeader ; 
			readHeader(mptr_FILE, mptr_ScanIndex[scan], &scanHeader) ;	

			if (strstr(scanHeader.msTypeString, "FTMS"))
				return true ; 
			else if (strstr(scanHeader.msTypeString, "ITMS"))
				return false ; 
		 
			//default
			return false ; 			
		}

		
		bool MZXmlRawData::IsProfileScan(int scan_num)
		{
			int error = -1 ; 
			struct ScanHeaderStruct scanHeader ; 
			readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &scanHeader) ;			
			
			if (scanHeader.profileType == 'p')
				return true ; 
			else if (scanHeader.profileType == 'c') 
				return false ; 
			
			//default
			return true ; 
		}

		
		int MZXmlRawData::GetMSLevel(int scan_num) 
		{
			 int ms_level = 0 ;
			 struct ScanHeaderStruct scanHeader ; 
			 readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &scanHeader) ;	
			 return scanHeader.msLevel ; 			 
		}

		double MZXmlRawData::GetParentMz(int scan_num)
		{
			double parent_mz = 0 ; 
			int ms_level = 0 ;
			struct ScanHeaderStruct scanHeader ; 
			readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &scanHeader) ;	
			return scanHeader.precursorMZ ; 		
		}
		
		bool MZXmlRawData::IsMSScan(int scan_num)
		{
			int ms_level = GetMSLevel(scan_num) ; 
			if (ms_level == 1)
				return true ; 
			else
				return false ; 
		}

		bool MZXmlRawData::GetRawData(std::vector <double>	*mzs, std::vector<double> *intensities, int scan_num) 
		{
			return GetRawData(mzs, intensities,	scan_num, -1) ;	
		}

		bool MZXmlRawData::GetRawData(std::vector <double>	*mzs, std::vector<double> *intensities, int scan_num, int num_pts)
		{

			if (scan_num > mint_num_scans)
			{
				char mesg[MAX_ERR_LEN] ; 
				strcpy(mesg, "File only has ") ; 
				_itoa(mint_num_scans, &mesg[strlen(mesg)], 10) ; 
				strcat(mesg, " scans. Cannot read to scan number: ") ; 
				_itoa(scan_num, &mesg[strlen(mesg)], 10) ; 
				throw mesg ; 
			}

			mint_current_scan = scan_num ; 
			readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &mobj_CurrentScanHeader);

			mzs->clear() ; 
			intensities->clear() ; 

			int num_points = mobj_CurrentScanHeader.peaksCount ; 
			if (num_pts > 0 && num_pts < num_points)
			{
				num_points = num_pts ; 
			}

			mzs->reserve(num_points) ; 
			intensities->reserve(num_points) ; 

			float *peaks = readPeaks(mptr_FILE, mptr_ScanIndex[mint_current_scan]);
			float fMass;
			float fInten;
			double max_intensity = -1*DBL_MAX, min_intensity = DBL_MAX; 
			for (int pt_num = 0 ; pt_num < 2*num_points-1 ; pt_num+=2)
			{	   
				fMass=peaks[pt_num];
				fInten=peaks[pt_num+1];
				if (fInten > max_intensity)
					max_intensity = fInten ; 
				if (fInten < min_intensity) //Anoop Jan 08: was > before
					min_intensity = fInten ; 
				mzs->push_back(fMass) ; 
				intensities->push_back(fInten) ; 
			}

			delete [] peaks ; 
			mdbl_signal_level = max_intensity - min_intensity ; 

			return true	; 
		}

		double MZXmlRawData::GetSignalRange(int scan_num)
		{
			return 0 ; 
		}

		void MZXmlRawData::GetTicFromFile(std::vector<double> *intensities, std::vector<double> *scan_times, bool base_peak_tic) 
		{
			if (mptr_FILE == NULL)
				return ; 

			intensities->clear() ; 
			scan_times->clear() ; 

			struct ScanHeaderStruct scanHeader ;
			bool nonzero_time_reported = false ; 
			for (int scan_num = 0 ; scan_num <= mint_num_scans ; scan_num++)
			{
				// its time to read in that scan.
				mint_current_scan = scan_num ; 
				readHeader(mptr_FILE, mptr_ScanIndex[scan_num], &scanHeader) ;
				scan_times->push_back(scanHeader.retentionTime) ; 
				if (scanHeader.retentionTime != 0.0)
					nonzero_time_reported = true ; 
				if (base_peak_tic)
				{
					intensities->push_back(scanHeader.basePeakIntensity) ; 
				}
				else
				{
					intensities->push_back(scanHeader.totIonCurrent) ; 
				}
			}

			// for mzxml files in which no retention times were recorded. 
			if (!nonzero_time_reported)
			{
				for (int scan_num = 0 ; scan_num <= mint_num_scans ; scan_num++)
				{
					(*scan_times)[scan_num] = scan_num ; 
				}
			}


		}

	}
}

File: C#_Version/DeconEngine/Readers/MZXmlRawData.cs
using System;
using System.Collections.Generic;
using DeconToolsV2.Readers;
using MSDataFileReader;

namespace Engine.Readers
{
    internal class MZXmlRawData : RawData
    {
        private string mstr_file_name;
        private int mint_num_scans;
        private int mint_current_scan;
        private double mdbl_signal_level;
        private MSDataFileReader.SpectrumInfo mobj_CurrentScanData;
        private MSDataFileReader.MzXMLFileAccessor mobj_MzXmlFile;

        public override short GetSpectrumType(int scan_num)
        {
            return 0;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override void GetScanDescription(int scan, out string description)
        {
            description = "Scan #" + scan;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override int GetNextScanNum(int current_scan_num)
        {
            return current_scan_num + 1;
        }

        public override int GetFirstScanNum()
        {
            return 1;
        }

        public override int GetLastScanNum()
        {
            return mint_num_scans;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override int GetNumScansLoaded()
        {
            return mint_current_scan;
        }

        public MZXmlRawData()
        {
            mint_num_scans = 0;
            mint_current_scan = -1;
            mobj_MzXmlFile = null;
        }

        ~MZXmlRawData()
        {
            if (mobj_MzXmlFile != null)
            {
                Close();
                mobj_MzXmlFile = null;
            }
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override string GetFileName()
        {
            return mstr_file_name;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override FileType GetFileType()
        {
            return FileType.MZXMLRAWDATA;
        }

        public override void Close()
        {
            mint_num_scans = 0;
            if (mobj_MzXmlFile != null)
            {
                mobj_MzXmlFile.CloseFile();
                mobj_MzXmlFile = null;
            }
        }

        public override int GetNumScans()
        {
            return mint_num_scans;
        }

        public override double GetScanTime(int scan_num)
        {
            if (mint_current_scan == scan_num)
            {
                return mobj_CurrentScanData.RetentionTimeMin;
            }
            else
            {
                var scanData = new SpectrumInfo();

                // its time to read in that scan.
                mint_current_scan = scan_num;
                mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out scanData);
                return scanData.RetentionTimeMin;
            }
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override int GetScanSize()
        {
            return 0;
        }

        public override void Load(string file_name)
        {
            mstr_file_name = file_name;
            Close();
            mobj_MzXmlFile = new MzXMLFileAccessor();
            mobj_MzXmlFile.OpenFile(mstr_file_name);
            mobj_MzXmlFile.ReadAndCacheEntireFile();
        }

        public override int GetParentScan(int scan_num)
        {
            var scan = scan_num;
            var scanData = new SpectrumInfo();
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan, out scanData);

            for (scan = scan_num - 1;; scan--)
            {
                mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan, out scanData);
                if (scanData.MSLevel == 1)
                    break;
            }

            return scan;
        }

        public override bool IsFTScan(int scan_num)
        {
            var scan = scan_num;
            var scanData = new SpectrumInfo();
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan, out scanData);
            mobj_MzXmlFile.GetSourceXMLByScanNumber(scan_num, out var xmlText);
            // TODO: Validate this solution
            if (xmlText.Contains("FTMS"))
            {
                return true;
            }
            else if (xmlText.Contains("ITMS"))
            {
                return false;
            }

            //default
            return false;
        }

        public override bool IsProfileScan(int scan_num)
        {
            var scanData = new SpectrumInfo();
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out scanData);
            // TODO: Validate this solution
            //if (scanData.profileType == 'p')
            //    return true;
            //else if (scanData.profileType == 'c')
            //    return false;
            return !scanData.Centroided;

            //default
            //return true;
        }

        public override int GetMSLevel(int scan_num)
        {
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out var scanData);
            return scanData.MSLevel;
        }

        public override double GetParentMz(int scan_num)
        {
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out var scanData);
            return scanData.ParentIonMZ;
        }

        public override bool IsMSScan(int scan_num)
        {
            var ms_level = GetMSLevel(scan_num);
            if (ms_level == 1)
                return true;
            else
                return false;
        }

        public override bool IsZoomScan(int scan_num)
        {
            return false;
        }

        // Note that Centroid is ignored by this class
        public override bool GetRawData(out List<double> mzs, out List<double> intensities, int scan_num, bool centroid)
        {
            return GetRawData(out mzs, out intensities, scan_num, centroid, -1);
        }

        // Note that Centroid is ignored by this class
        public override bool GetRawData(out List<double> mzs, out List<double> intensities, int scan_num, bool centroid, int num_pts)
        {
            if (scan_num > mint_num_scans)
            {
                var mesg = "File only has " + mint_num_scans + " scans. Cannot read to scan number: " + scan_num;
                throw new Exception(mesg);
            }

            mint_current_scan = scan_num;
            mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out mobj_CurrentScanData);

            mzs = new List<double>();
            intensities = new List<double>();

            var scanData = new SpectrumInfo();
            mobj_MzXmlFile.GetSpectrumByScanNumber(scan_num, out scanData);

            var num_points = scanData.DataCount;
            if (num_pts > 0 && num_pts < num_points)
            {
                num_points = num_pts;
            }

            mzs.Capacity = num_points;
            intensities.Capacity = num_points;
            double fMass;
            double fInten;
            var max_intensity = -1 * double.MaxValue;
            var min_intensity = double.MaxValue;
            for (var i = 0; i < scanData.DataCount; i++)
            {
                fMass = scanData.MzList[i];
                fInten = scanData.IntensityList[i];
                if (fInten > max_intensity)
                    max_intensity = fInten;
                if (fInten < min_intensity) //Anoop Jan 08: was > before
                    min_intensity = fInten;
                mzs.Add(fMass);
                intensities.Add(fInten);
            }

            mdbl_signal_level = max_intensity - min_intensity;

            return true;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override double GetSignalRange(int scan_num, bool centroid)
        {
            return 0;
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public override void GetTicFromFile(out List<double> intensities, out List<double> scan_times,
            bool base_peak_tic)
        {
            intensities = new List<double>();
            scan_times = new List<double>();

            if (mobj_MzXmlFile == null)
                return;

            var nonzero_time_reported = false;
            for (var scan_num = 0; scan_num <= mint_num_scans; scan_num++)
            {
                // its time to read in that scan.
                mint_current_scan = scan_num;
                var scanData = new SpectrumInfo();
                mobj_MzXmlFile.GetSpectrumHeaderInfoByScanNumber(scan_num, out scanData);
                scan_times.Add(scanData.RetentionTimeMin);
                if (scanData.RetentionTimeMin != 0.0)
                    nonzero_time_reported = true;
                if (base_peak_tic)
                {
                    intensities.Add(scanData.BasePeakIntensity);
                }
                else
                {
                    intensities.Add(scanData.TotalIonCurrent);
                }
            }

            // for mzxml files in which no retention times were recorded.
            if (!nonzero_time_reported)
            {
                for (var scan_num = 0; scan_num <= mint_num_scans; scan_num++)
                {
                    scan_times[scan_num] = scan_num;
                }
            }
        }
    }
}
